File: contests/lc/1.TwoSum/two_sum.go
package __TwoSum

func twoSum(nums []int, target int) []int {
	numToIndex := make(map[int]int)

	for i, num := range nums {
		if j, ok := numToIndex[target-num]; ok {
			return []int{j, i}
		}
		numToIndex[num] = i
	}

	return nil // по условию задачи такого не будет
}

File: contests/lc/1.TwoSum/two_sum.py
from typing import List


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # Track indices of seen numbers to find complements in O(n)
        indices_by_value = {}
        for idx, value in enumerate(nums):
            complement = target - value
            if complement in indices_by_value:
                return [indices_by_value[complement], idx]
            indices_by_value[value] = idx
        return []

    def twoSum2(self, nums: List[int],target: int) -> List[int]:
        d = {}
        for i,x in enumerate(nums):
            if (y := target -x) in d:
                return [d[y],i]
            d[x] =i


def main() -> None:
    nums = [2, 7, 11, 15]
    target = 9
    result = Solution().twoSum(nums, target)
    result2 = Solution().twoSum2(nums,target)
    print(result,result2)


if __name__ == "__main__":
    main()
